### PR TITLE
E2E test cases for Invalid user inputs for vSphere

### PR DESCRIPTION
**Pull request by BaluDontu**
Add E2E test cases to invalid inputs for vSphere

- Invalid FS Type in Storage Class.
- Invalid datastoreName in Storage Class.
- Invalid disksize for a persistent volume.

@divyenpatel @rohitjogvmw

### CODE REVIEW

**Comment by rootfs** [comment]
I prefer to build these test cases into vsphere volume unit or integration tests. Testing in an e2e test setup and validating the results through e.g. error message is fragile.

File: test/e2e/storage/vsphere_volume_datastore.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		err := invokeInvalidDatastoreTestNeg(client, namespace, scParameters)
+		Expect(err).To(HaveOccurred())
+		errorMsg := `Failed to provision volume with StorageClass \"` + DatastoreSCName + `\": datastore '` + InvalidDatastore + `' not found`
+		if !strings.Contains(err.Error(), errorMsg) {
+			Expect(err).NotTo(HaveOccurred(), errorMsg)
+		}
+	})
+})
+
+func invokeInvalidDatastoreTestNeg(client clientset.Interface, namespace string, scParameters map[string]string) error {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_datastore.go]: Not sure `error` is the right return value here. There are several calls to `Expect` so `error` is not returned in these cases. At the end `error` is a list of events.

File: test/e2e/storage/vsphere_volume_disksize.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		err := invokeInvalidDiskSizeTestNeg(client, namespace, scParameters, diskSize)
+		Expect(err).To(HaveOccurred())
+		errorMsg := `Failed to provision volume with StorageClass \"` + DiskSizeSCName + `\": A specified parameter was not correct`
+		if !strings.Contains(err.Error(), errorMsg) {
+			Expect(err).NotTo(HaveOccurred(), errorMsg)
+		}
+	})
+})
+
+func invokeInvalidDiskSizeTestNeg(client clientset.Interface, namespace string, scParameters map[string]string, diskSize string) error {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_disksize.go]: same comments on returning `error` as above.

File: test/e2e/storage/vsphere_volume_fstype.go [hunk 3]
```diff
@@ -17,17 +17,26 @@ limitations under the License.
...
 	"time"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	k8stype "k8s.io/apimachinery/pkg/types"
 	clientset "k8s.io/client-go/kubernetes"
-	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere"
+	vsphere "k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere"
```
Comment (jeffvance) [inline test/e2e/storage/vsphere_volume_fstype.go]: why is the "vsphere" alias needed?

**Comment by jeffvance** [comment]
I agree with @rootfs that these new tests seem more like unit tests than e2e.

**Comment by BaluDontu** [comment]
@jeffvance : Till now we have only covered positive test cases as part of E2E test framework. Adding more negative test cases to E2E test suite will provide more stability.
Atleast in case of vSphere, we have observed these issues running more often. So adding these very minimal test cases for basic invalidation would be more useful.

**Comment by jeffvance** [comment]
@BaluDontu  I understand your point. Can you provide some insights as to why the vsphere unit tests did/do not catch erogenous size and fs type errors?

**Comment by BaluDontu** [comment]
@jeffvance : Unit tests only contain mock implementation. We want atleast these very basic negative test cases that are included in this PR to be executed against a real vSphere environment.

**Comment by jeffvance** [comment]
@BaluDontu I think someone with more experience in the design philosophy concerning the different types of tests should weigh in on this pr. I applaud having more tests to discover bugs before a customer finds the bug. At the same time I am aware of issues related to unowned, flaky e2e tests that block the submit queue and take up a lot of time. I understand that the vsphere tests are owned but I still think we need someone with more test history to take a look at your pr. @jayunit100 @timstclair

**Comment by BaluDontu** [comment]
@jayunit100 @timstclair any thoughts on this ?

**Comment by BaluDontu** [comment]
@jeffvance : I haven't heard back from @jayunit100 @timstclair. Is there any way that we can expedite the PR review process.

**Comment by jayunit100** [comment]
looking now

**Comment by tallclair** [comment]
s/ @timstclair / @timothysc / ?

**Comment by jayunit100** [comment]
Currently discussing in @kubernetes/sig-testing-misc  wether or not more VSPhere should tests should be flowing in here...  i dont know the current state of test scope for e2e, i do know we're trying to prune them to very well maintained ones only , though,

**Comment by jeffvance** [comment]
/ok-to-test
/lgtm
Since no comment from @jayunit100 or @timstclair  I will err on the side of more tests are better and
/approve

**Comment by jayunit100** [comment]
@jeffvance lets hold off on the approval

**Comment by jayunit100** [comment]
/do-not-merge/release-note-label-needed

**Comment by jeffvance** [comment]
Sorry I had the wrong name for tim st clair...

**Comment by spiffxp** [comment]
I would at least like to see a `[Feature:VSphereStorage]` or something to that effect in the test names so they don't get run by default

**Comment by BaluDontu** [comment]
@spiffxp : If you see all the vsphere tests, you will see the following piece of code
framework.SkipUnlessProviderIs("vsphere")
which says to skip all the tests to run unless the provider is vsphere. 

@jeffvance : Based on offline discussion in sig-testing slack channel, I think we should be good to proceed with this change. Let me know if there are any other queries.

**Comment by jeffvance** [comment]
I know the test is skipped unless the provisioner is vsphere, but do you want to address @spiffxp 's comment re. adding a `[Feature:vsphere]` tag?  If this is agreed to then I would assume it's ok to be in a separate pr since all vsphere tests would need this tag, and other storage tests would need to have a Feature tag added too (I assume).

**Comment by BaluDontu** [comment]
@jeffvance : Can you please take a look at the latest code. I have [Feature:vsphere] tag to the tests for the files included in this PR.

File: test/e2e/storage/vsphere_volume_fstype.go (hunk 3 above)
Comment (jeffvance) [reply test/e2e/storage/vsphere_volume_fstype.go]: ping @BaluDontu

**Comment by jeffvance** [comment]
other than a minor comment-question, it looks good to me.
/lgtm

**Comment by cjwagner** [comment]
> jingxu97 removed the do-not-merge/release-note-label-needed label 12 minutes ago

/release-note-none

**Comment by cblecker** [comment]
/lgtm cancel
Removing LGTM to stop re-testing loop. Test is not building:
```
W1007 02:09:33.581] # k8s.io/kubernetes/test/e2e/storage
W1007 02:09:33.582] test/e2e/storage/vsphere_volume_vsan_policy.go:336: not enough arguments in call to getVSphereClaimSpecWithStorageClassAnnotation
W1007 02:09:33.582] 	have (string, *"k8s.io/kubernetes/vendor/k8s.io/api/storage/v1".StorageClass)
W1007 02:09:33.582] 	want (string, string, *"k8s.io/kubernetes/vendor/k8s.io/api/storage/v1".StorageClass)
```

Looks like it might need a rebase as #53227 was merged and changed this file.

**Comment by BaluDontu** [comment]
I have rebased this PR to resolve the conflicts. 
@jingxu97 please take a look at it.

**Comment by timothysc** [comment]
I'd defer to @kubernetes/sig-storage-pr-reviews on this one due to ownership.

**Comment by BaluDontu** [comment]
@jingxu97 @jeffvance : i have to rebase the PR again for some conflicts. Please take a look and provide an lgtm if everything looks good.

**Pull request merged**